README.md
UniPath
=======

The library for universal access to any data from a PHP script (similar to xpath with jquery philosophy).

Current version: [unipath-2.4rc3.php](https://github.com/SaemonZixel/unipath.php/raw/master/unipath-2.4.php)

Old version: [unipath-2.2.3.php](https://github.com/SaemonZixel/unipath.php/raw/master/unipath.php)

DataBase
--------

	global $db1; // for 5.3 and upper

	$db1 = new PDO(...);
	
	uni('/db1/table1[]') 
	// array(
	//  0 => array('id' => 1, 'field1' => 123, 'field2' => 'abc' ),
	//  1 => array('id' => 2, 'field1' => 456, 'field2' => 'abc')
	// )
	
	uni('/db1/table1[]/0')
	// array('id' => 1, 'field1' => 123, 'field2' => 'abc')
	
	uni('/db1/table1[]/1')
	// array('id' => 2, 'field1' => 456, 'field2' => 'abc')
	
	uni('/db1/table1[]/first()')
	// array('id' => 1, 'field1' => 123, 'field2' => 'abc')
	
	uni('/db1/table1[id = 1]') 
	// array(0 => array('id' => 1, 'field1' => 123, 'field2' => 'abc'))
	
	uni("/db1/table1[id = 1 and field2 = 'abc']")
	// array(0 => array('id' => 1, 'field1' => 123, 'field2' => 'abc'))

	uni("/db1/table1[field2 = 'abc']")
	// array(
	//  0 => array('id' => 1, 'field1' => 123, 'field2' => 'abc'),
	//  1 => array('id' => 2, 'field1' => 456, 'field2' => 'abc')
	// )
	
	uni("/db1/table1[field2 = 'abc']/1")
	// array('id' => 1, 'field1' => 456, 'field2' => 'abc')

	uni("/db1/table1[field2 = 'abc']/1/id")
	// 1

	uni("/db1/table1[field1 >= 400 and field2 < 700 and id = 1,2,3]/0")
	// array('id' => 1, 'field1' => 456, 'field2' => 'abc')

	uni("/db1/table1/id")
	// array(1, 2)

	uni("/db1/table1[table1.id = 1,2]+table2[table1.id = table2.ref_id])")
	// array(
	//   0 => array('id' => 1, 'field1' => 123, 'field2' => 'abc', 'ref_id' => 1, 'field_from_table2' => 1.23),
	//   1 => array('id' => 2, 'field1' => 456, 'field2' => 'abc', 'ref_id' => 2, 'field_from_table2' => 4.56)
	// )

	uni("/db1/table1[table1.id = 1,2]+table2[table1.id = table2.ref_id]/order_by('table1.field1 DESC')/columns('table1.id, field_from_table2, table2.id AS id2')/limit(0, 2)")
	// array(
	//   0 => array('id' => 2, 'field_from_table2' => 4.56, 'id2' => 102),
	//   1 => array('id' => 1, 'field_from_table2' => 1.23, 'id2' => 101)
	// )

	uni("/db1/table1[table1.id = 1,2]+table2[table1.id = table2.ref_id]/order_by('table1.field1 DESC')/columns('table1.id, field_from_table2, table2.id AS id2, ifNULL(field1, 999) as field1')/limit(0, 2)/asSQLQuery()")
	// array(
	//   'sql_query' => 'SELECT table1.id, field_from_table2, table2.id AS id2, ifNULL(field1, 999) as field1 FROM table1 LEFT JOIN table2 ON table1.id = table2.ref_id WHERE table1.id IN (1,2) ORDER BY table1.field1 DESC LIMIT 0, 2',
	//   'sql_params' => array()
	// )
	
	uni("/db1/alias('table1', 't1')[t1.id = 1,2]+table2-as-t2[t1.id = t2.ref_id]/asSQLQuery()")
	// array(
	//   'sql_query' => 'SELECT * FROM table1 AS t1 LEFT JOIN table2 AS t2 ON t1.id = t2.ref_id WHERE t1.id IN (1,2)',
	//   'sql_params' => array()
	// )
	// table_name-as-alias_name is syntactical sugar for alias() function
	
	uni("/db1/sql_query('SELECT * FROM table1')/0")
	// array('id' => 1, 'field1' => 456, 'field2' => 'abc')
	
	uni("/db1/table1[]/toHash('field1')")
	// array(
	//  123 => array('id' => 1, 'field1' => 123, 'field2' => 'abc'),
	//  456 => array('id' => 2, 'field1' => 456, 'field2' => 'abc')
	// )
	
	Bracket grouping of the WHERE-clauses is supported partially yet.
	
	
DataBase (UPDATE, INSERT, DELETE)
---------------------------------

	Warning: the beta version of the library may contain bugs, that can delete or damage your data in the database.

	Supported database drivers: ODBC, PDO, mysql_*, mysqli_*.
	
	For MS SQL Server there are special functions top(), sql_iconv(), like2(), chunked(), which have not been described yet.

	uni("/db1/table1[id = 1]", array('field1' => 999))
	// array('field1' => 999)
	
	uni("/db1/table1[id = 2]/0/field1", 888)
	// 888
	
	uni('/db1/table1[]')
	// array(
	//  0 => array('id' => 1, 'field1' => 999, 'field2' => 'abc'),
	//  1 => array('id' => 2, 'field1' => 888, 'field2' => 'abc')
	// )
	
	uni('/db1/table1[field2 = `abc`]', array('field1' => 123))
	// array('field1' => 123)
	
	uni('/db1/table1[]'); 
	// array(
	//  0 => array('id' => 1, 'field1' => 123, 'field2' => 'abc'),
	//  1 => array('id' => 2, 'field1' => 123, 'field2' => 'abc')
	// )
	
	uni('/db1/table1[]/1', array('field1' => 456))
	// array('field1' => 456)
	
	uni('/db1/table1[]/2', array('field1' => 789))
	// array('field1' => 789)
	//
	// Notice: the uni() will do a silent assignment! no Notice, no Warning!
	
	uni("/db1/table1/new_row()", array('id' => 3, 'field1' => 999, 'field2' => 'xxx'))
	// array('id' => 3, 'field1' => 999, 'field2' => 'xxx')
	
	uni("/db1/last_error()")
	// NULL
	
	uni("/db1/last_affected_rows()")
	// 1
	
	uni("/db1/last_insert_id()")
	// 3
	
	uni("/db1/table1[id = 3]/delete()")
	// array()
	
	uni("/db1/last_affected_rows()")
	// 1
	
Array
-----
	
	global $array1; // for 5.3 and upper

	$array1 = array('aaa' => 111, 'bbb' => 222, 'ccc' => 333)
	
	uni("/array1")
	// array('aaa' => 111, 'bbb' => 222, 'ccc' => 333)
	
	uni("/array1/aaa")
	// 111
	
	uni("/array1/xxx")
	// null
	
	uni("/array1/xxx/ifEmpty(999)")
	// 999
	
	uni("/array1/count()")
	// 3
	
	uni("/array1/.")
	// array('aaa' => 111, 'bbb' => 222, 'ccc' => 333)
	//
	// Notice: dot "." works like self() in XPath
	
	uni("/array1/./././././.")
	// array('aaa' => 111, 'bbb' => 222, 'ccc' => 333);
	
	uni("/array1/.[. > 200]")
	// array('bbb' => 222, 'ccc' => 333)
	
	uni("/array1/.[. > 200 and . < 300]")
	// array('bbb' => 222)
	
	uni("/array1/aab", 112)
	// 112
	
	uni("/array1/aa%s")
	// array('aaa' => 111, 'aab' => 112)
	
	uni("/array1/aa%s[. > 111]")
	// array('aab' => 112)
	
	// condition: to skip or not
	uni("/array1/[aaa > 200]")
	// null
	
Objects
-------
	global $mysqli_obj; // for 5.3 and upper
	
	$mysqli_obj = uni("/MySQLi/MySQLi('127.0.0.1', 'user', 'password', 'database1')")
	// object(mysqli)#1 { ... }
	
	uni("/mysqli_obj/host_info")
	// "127.0.0.1 via TCP/IP"
	
	uni("/mysqli_obj/query('SELECT * FROM `users` WHERE `user_banned` = 1')")
	// object(mysqli_result) {
	//		'current_field' => 0,
    //		'field_count' => 3,
	//		'lengths' => null,
    //		'num_rows' => 129006,
    //		'type' => 0,
    //	}

    uni("/mysqli_obj/query(```SELECT * FROM `users` WHERE `user_banned` = 1```)/fetch_assoc()")
    // array(
	//	'user_id' => 1,
	//	'user_name' => 'test',
	//	'user_banned' => 1
    // )

	
XML
---

	global $xml; // for 5.3 and upper

	$xml = '<rows>
		<row id="1"><cell index="1">111_AAA</cell><cell index="2">111_BBB</cell></row>
		<row id="2"><cell index="1">222_AAA</cell><cell index="2">222_BBB</cell></row>
	</rows>';
	uni("/xml/asXML()")
	// object(DOMElement)#7 (18) {
	//    ["nodeName"]=> 
	//    string(4) "rows"
	//    ["nodeValue"]=>
	//    string(16) "
	//        111_AAA111_BBB
	//        222_AAA222_BBB
	//    "
	//    ...
	// }
	
	uni("/xml/asXML()/nodeName")
	// 'rows'
	
	uni("/xml/asXML()/row")
	// array(
	//    0 =>
	//    object(DOMElement)#7 (18) {
	//       ["nodeName"]=>
	//       string(3) "row"
	//       ...
	//    },
	//    1 =>
	//    object(DOMElement)#3 (18) {
    //       ["nodeName"]=>
    //       string(3) "row"
	//       ...
	//    }
	// )
	
	
	uni("/xml/asXML()row/0")
	// object(DOMElement)#7 (18) {
	//    ["nodeName"]=> 
	//    string(4) "row"
	//    ["nodeValue"]=>
	//    string(16) "111_AAA111_BBB"
	//    ...
	// }
	
	
	uni("/xml/asXML()/row/0/cell")
	// array(
	//    0 =>
	//    object(DOMElement)#9 (18) {
	//       ["nodeName"]=>
	//       string(3) "cell"
	//       ["nodeValue"]=>
    //       string(2) "111_AAA"
	//       ...
	//    },
	//    1 =>
	//    object(DOMElement)#11 (18) {
    //       ["nodeName"]=>
    //       string(3) "cell"
	//       ["nodeValue"]=>
    //       string(2) "111_BBB"
    //       ...
	//    }
	// )
	
	uni("/xml/asXML()/row/cell")
	// array(
	//    0 =>
	//    object(DOMElement)#9 (18) {
	//       ["nodeName"]=>
	//       string(3) "cell"
	//       ["nodeValue"]=>
    //       string(2) "111_AAA"
	//       ...
	//    },
	//    1 =>
	//    object(DOMElement)#11 (18) {
    //       ["nodeName"]=>
    //       string(3) "cell"
	//       ["nodeValue"]=>
    //       string(2) "111_BBB"
    //       ...
	//    },
	//    2 =>
	//    object(DOMElement)#14 (18) {
    //       ["nodeName"]=>
    //       string(3) "cell"
	//       ["nodeValue"]=>
    //       string(2) "222_AAA"
    //       ...
	//    },
	//    3 =>
	//    object(DOMElement)#17 (18) {
    //       ["nodeName"]=>
    //       string(3) "cell"
	//       ["nodeValue"]=>
    //       string(2) "222_BBB"
    //       ...
	//    }
	// )
	
	uni("/xml/asXML()/row/0/cell/0/nodeValue")
	// '111_AAA'
	
	uni("/xml/asXML()/row/0/getAttribute('id')")
	// '1'
	
	/* syntactical suggar to get attribute value more shortly */
	uni("/xml/asXML()/row/0/@id")
	// '1'
	
	uni("/xml/asXML()/row/*/@id")
	// array('1', '2')
	
	uni("/xml/asXML()/row/cell/*/nodeValue")
	// array('111_AAA', '111_BBB', '222_AAA', '222_BBB')
	
	global $array1;
	$array1 = array('items' => array('aaa' => 111, 'bbb' => 222, 'ccc' => 333));
	uni('/array1/ArrayToXML()')
	// '<items><aaa>111</aaa><bbb>222</bbb><ccc>333</ccc></items>'
	
Files
-----

	uni("/fs()/etc/passwd/contents()")
	// ... contents of the 'passwd' file ...
	
	$filename = '/etc/passwd';
	uni("/filename/asFile()/contents()")
	// ... contents of the 'passwd' file ...
	
	$dirname = '/etc';
	uni("/dirname/asDirectory()/passwd/contents()")
	// ... contents of the 'passwd' file ...
	
Image
-----

	uni("/fs()/images/image1.jpg/asImageFile()/resize(400, 500, 'fill')/crop(400, 500)/saveAs('/images/image1b.png')")
	// (resized and croped image saved as '/images/image1b.png')
	
class Uni
---------

Class Uni extends ArrayIterator and works like Array.

	$rows = new Uni('/db1/table1[]')
	// object(Uni)
	
	$rows[0]
	// array('id' => 1, 'field1' => 123, 'field2' => 'abc')
	
	$rows[0] = array('field1' => 0)
	// array('field1' => 0)
	
	$rows[0]
	// array('id' => 1, 'field1' => 0, 'field2' => 'abc')
	
	$rows['0/field2'] = 'ddd'
	// 'ddd'
	
	$rows[0]
	// array('id' => 1, 'field1' => 0, 'field2' => 'ddd')
	
	$rows['0/field1,field2'] = array('field1' => 111, 'field1' => 'ccc');
	// array('field1' => 111, 'field1' => 'ccc')
	
	$rows[0]
	// array('id' => 1, 'field1' => 111, 'field2' => 'ccc')
	
	$rows['*/field2'] = 'eee';
	// 'eee'
	
	$rows['all()']
	// array(
	//  0 => array('id' => 1, 'field1' => 111, 'field2' => 'eee' ),
	//  1 => array('id' => 2, 'field1' => 456, 'field2' => 'eee')
	// )
	
	foreach($rows as $row) var_dump($row);
	// array('id' => 1, 'field1' => 111, 'field2' => 'ccc')
	// array('id' => 2, 'field1' => 123, 'field2' => 'abc')
	
Cache
-----

	uni('cached(/cached_var1, /db1/table1[id=1])/0')
	// array('id' => 1, 'field1' => 111, 'field2' => 'ccc');
	//
	// You can write cached($cached_var1, ...) it is equal
	
	uni('cached(/cached_var1)/0/id')
	// 1

	var_dump($GLOBALS['cached_var1'])
	// array('id' => 1, 'field1' => 111, 'field2' => 'ccc');
	
	uni('/cached_var1/0/id')
	// 1
	
Templating
----------
	
	uni('/db1/table1[]/ssprintf(`<option value="%s"%s>%s</option>`, id, if([id=2], ` selected`, ``), field2)/join(`\n`)')
	// <option value="1">ccc</option>
	// <option value="2" selected>abc</option>
	
Debugging
---------

	$GLOBALS['unipath_debug_sql'] = true; // show all sql queres
	
	$GLOBALS['unipath_debug'] = true; // activate debugging of execution process
	
	$GLOBALS['unipath_debug_parse'] = true; // activate debugging of parsing process
	
	$GLOBALS['unipath_error_reporting'] = E_ALL; // change internal error_reporting level
	
	global $__uni_prt_cnt;
	$__uni_prt_cnt = 1000000; // change infinity loop protection limit
	

ChangeLog.txt
2021-12-00  Saemon Zixel  <saemonzixel@gmail.com>

	* version 2.4rc4
	* image: toJPEG(), asImage()
	* bug fixes

2021-02-18  Saemon Zixel  <saemonzixel@gmail.com>
	
	* version 2.4rc3
	* added $unipath_error_reporting option
	* NameSpace support added
	* XML: finish refactoring
	* bug fixes

2020-12-06  Saemon Zixel  <saemonzixel@gmail.com>
	
	* version 2.4rc2
	* class UniPathExtension
	* SQL: mysqli support added 
	* SQL: table_name-as-alias_name sintactic sugar
	* SQL: improvements in UPDATE, DELETE
	* brackets in expression (only for or/and)
	* XML: refactored halfly (and broken halfly)
	* added asClass(), asZip()
	* many bug fixes
	* unipath.email: bodyHtml(), subject, body fields access

2019-02-20  Saemon Zixel  <saemonzixel@gmail.com>
	
	* version 2.2.3: now compatible with php 7.x
	* php-foreach: added
	* many bug fixes
	
2018-02-13  Saemon Zixel  <saemonzixel@gmail.com>

	* unipath.email.php: extension is added for sending emails.
	* unipath.captcha.php: extension is added for generating captcha.

2017-02-15  Saemon Zixel  <saemonzixel@gmail.com>
	
	* version 2.2.2 
	* SQL: fixed an issue with float in SQL WHERE clouse 
	* /CONST added
	* toArray() repaired
	
2016-09-02  Saemon Zixel  <saemonzixel@gmail.com>

	* version 2.2.1-beta
	* repaired parsing `, ``` and ``````` in unipath

2016-07-12  Saemon Zixel  <saemonzixel@gmail.com>
	
	* version 2.2-dev
	* now uni() return array() and Uni[] return object
	* SQL: sql_table_prefix(), delete(), new() alias of new_row()
	* php:<func_name>() a new type of function call with manually specifying the arguments
	* Uni::<uni_func>() now accept many arguments
	* /[...]/ pass or not to pass
	* cached(var_name, unipath_to_cache) - new style of function cache()
	* if(<unipath>, <if_true>, <if_false>), regexp(), wrap() - can accept right as second argument, translit() - now invertible, untranslit() - invert translit() result, sprintf(), sprintf1()
	* some common variables were globalized
	* some bug fixes and new bugs

2016-05-09  Saemon Zixel  <saemonzixel@gmail.com>

	* version 2.1-dev
	* SQL: sql_result_cache(...) for reiterating, NULL support on update
	* OOP: Class::<static_method>(), Class::<__construct | Class>() -> new Class(), Object-><property>, Object-><method>(arg1, ...)
	* next_with_key(), next() don`t accept arguments now!
	* unipath_tests.php: auto search extensions and launching tests from them
	* unipath.linedoc.php: the extension for to parse LineDocument format
	* some bug fixes

2016-04-18  Saemon Zixel  <saemonzixel@gmail.com>
	
	* version 2.0.1-dev
	* class Uni: refactored and repaired, add Uni::<func_name>(args) for an easy call of _uni_<func_name>()
	* SQL: table1++table2 -> table1 LEFT OUTER JOIN table2
	* bug fixes in resize()
	* next(arg1, arg2) - returns array(arg1, arg2) for the next data element
	* same bug fixes and new tests

2016-04-08  Saemon Zixel  <saemonzixel@gmail.com>

	* version 2.0-dev
	* cursor() - a new interface for the extensions. All the extensions/functions changed to work as the cursor().
	* __uni_parseFunc(): parses the arguments (example: /array(id = 1, value1 = 'test'))
	* .../...%s... - sscanf can be used to filter by a key
	* __uni_parseFunc(): `, ```, ``````` - is beeing recognized as the beginning and the end of the string  
	* SQL: now data in database can be changed using uni()!
	* SQL: new_row() - for INSERT queries, last_error(), last_insert_id(), last_affected_rows(), group_by() - for GROUP BY clause
	* SQL: support mysql_* added
	* class Uni(): for OOP style programming. Implements ArrayIterator interface! But can be buggy :(
	* ZIP: asZipFile() - for reading the zip-archives, .../NNN/contents(), .../NNN/saveAs()
	* new functions: first(), next(), all(), ifNull(), url(), open(), replace() - to map array items, array() - to create arrays on the fly, key(), key(name%i), normalize_float()
	* deleted $uni_cache_data. Now cache() saves data in $GLOBALS!
	* trigger_error() - used for all internal errors and warnings
	* $__uni_optimeze - default is 1 (to optimize simple unipaths)
	* $GLOBALS['unipath_benchmark']: true - for internal measuring execution time
	* many bug fixes, tests and new bugs...
	
2016-02-07  Saemon Zixel  <saemonzixel@gmail.com>

	* version 1.7.8dev
    * SQL: an empty filter or with a half expression
    * set(./, ./), toURLTranslit(), url_seg()
    * strings in the unipath segments with delimiters: ', ", `, ```.
    * bug fixes and new tests

2015-07-05 Saemon Zixel  <saemonzixel@gmail.com>

	* version 1.7.7dev
	* SQL: ilike()
	* Image: asImageFile(), resize(), crop(), watermark(), saveAs()
	* fs(), basename(), url(), url()/contents() use sockets, split()+arrays, array_flat(), plus_procents()
	* secondary filtering can use a key() in the filter and a list of strings in the right hand expression
	* many bug fixes and new bugs...
    
2015-04-10 Saemon Zixel  <saemonzixel@gmail.com>

	* assignment is repaired, but little not very functional.
    * asXML(), ArrayToXML(), decode_url()
    * many bug fixes and new bugs
    
2015-02-16 Saemon Zixel  <saemonzixel@gmail.com>
	
	* unipath.php: many changes...
	* unipath_tests.php: tests for parsing and evaluаting
    
2015-01-22 Saemon Zixel  <saemonzixel@gmail.com>
	
	* RADME.md: DataBase and Array examples
    
2014-11-11 Saemon Zixel  <saemonzixel@gmail.com>
	
	* version 1.4-alpha
	* unipath.php: basic functionality

unipath.email.php
<?php
/**
 * E-mail and MIME extension for UniPath 2.2+
 * 
 * @version 1.1
 * @author Saemon Zixel (saemonzixel@gmail.com)
 * @license Public Domain
 *
 */

if(function_exists('_uni_email') == false) {

function _uni_asMIME() { 
	trigger_error("Not emplemented yet!", E_USER_ERROR); 
}

function _uni_email($tree, $lv = 0) {

	$result = array('data' => array(), 'data_type' => 'array/email', 
		'metadata' => array(
			'array/email', 
			'cursor()' => '_cursor_email',
			'boundary_separator' => md5(time())));
	
	list($args, $args_types) = __uni_parseFuncArgs($tree[$lv]['name']);
	foreach($args_types as $name => $type)
		if($type == 'unipath') {
			$uni_result = __uni_with_start_data(
				$tree[$lv-1]['data'], 
				isset($tree[$lv-1]['metadata']) 
					? $tree[$lv-1]['metadata'][0] 
					: $tree[$lv-1]['data_type'], 
				isset($tree[$lv-1]['data_tracking']) 
					? $tree[$lv-1]['data_tracking'] 
					: isset($tree[$lv-1]['metadata']) 
					? $tree[$lv-1]['metadata']
					: null, 
				$args[$name]);
			$result['data'][$name] = $uni_result['data'];
		} 
		else 
			$result['data'][$name] = $args[$name];

	return $result;
}

function _cursor_email($tree, $lv = 0, $cursor_cmd = null, $cursor_arg1 = null) {
// if($cursor_cmd != 'next')
// var_dump($tree[$lv]['name']." -- ".__FUNCTION__.".$cursor_cmd ".(is_array($cursor_arg1)&&isset($cursor_arg1['name'])?$cursor_arg1['name']:''));

	// EVAL resultHeaders()
	if($cursor_cmd == 'eval' and isset($cursor_arg1['name'][13]) and substr_compare($cursor_arg1['name'], "resultHeaders(", 0, 14, true) == 0) {
		
		$boundary_separator = $tree[$lv]['metadata']['boundary_separator'];
		$eol = "\r\n"; // PHP_EOL;

		$from = $tree[$lv]['data']['from'][0];
		$from = is_array($from) && count($from) > 1 
			? '=?UTF-8?B?'.base64_encode($from[1]).'?= <'.$from[0].'>'
			: $from[0];
		$headers = "From: ".$from . $eol;
			
		if(!empty($tree[$lv]['data']['bcc'])) {
			$bcc = $tree[$lv]['data']['bcc'][0];
			$bcc = is_array($bcc) && count($bcc) > 1 
				? '=?UTF-8?B?'.base64_encode($bcc[1]).'?= <'.$bcc[0].'>'
				: $bcc[0];
			$headers .= "Bcc: " . $bcc . $eol;
		}
		
		// заголовок в формате multipart
		$headers .= "MIME-Version: 1.0" . $eol;
		$headers .= "Content-Type: multipart/mixed; boundary=\"$boundary_separator\"";
		
		return array(
			'data' => $headers,
			'data_type' => 'string',
			'metadata' => array('string'));
	}
	
	// EVAL resultMessage()
	if($cursor_cmd == 'eval' and isset($cursor_arg1['name'][13]) and substr_compare($cursor_arg1['name'], "resultMessage(", 0, 14, true) == 0) {
	
		$boundary_separator = $tree[$lv]['metadata']['boundary_separator'];
		$eol = "\r\n"; // PHP_EOL;
	
		// предупреждение о MIME encoded message
		$body = "Content-Transfer-Encoding: 7bit" . $eol . $eol;
		$body .= "This is a MIME encoded message." . $eol . $eol;

		// тело сообщения
		// TODO content-type?
		if(!empty($tree[$lv]['data']['body'])) {
			$body .= "--" . $boundary_separator . $eol;
			$body .= "Content-Type: text/html; charset=\"utf-8\"" . $eol;
			$body .= "Content-Transfer-Encoding: 8bit" . $eol . $eol;
			$body .= $tree[$lv]['data']['body'] . $eol;
		}

		// прикррепляем файлы
		if(!empty($tree[$lv]['data']['files'])) 
		foreach($tree[$lv]['data']['files'] as $file_rec) {
// 			$file_size = filesize($file);
			$human_filename = isset($file_rec[1]) ? $file_rec[1] : $file_rec[0];
			$content = chunk_split(base64_encode(file_get_contents($file_rec[0])));
			if(!empty($content)) {
				$body .= "--" . $boundary_separator . $eol;
				$body .= "Content-Type: application/octet-stream; name=\"$human_filename\"".$eol;
				$body .= "Content-Transfer-Encoding: base64" . $eol;
				$body .= "Content-Disposition: attachment; filename=\"$human_filename\"" . $eol . $eol;
				$body .= $content . $eol;
			}
			else
				trigger_error("File {$file_rec[0]} is empty!", E_USER_ERROR);
		}
		$body .= "--" . $boundary_separator . "--";
		
		return array(
			'data' => $body,
			'data_type' => 'string',
			'metadata' => array('string'));
	}

	// EVAL xxx()
	if($cursor_cmd == 'eval' and $pos = strpos($cursor_arg1['name'], "(")) {
	
		// разберём аргументы
		list($args, $args_types) = __uni_parseFuncArgs($cursor_arg1['name']);
// 		assert('empty($args["arg1"]) == false;');
		foreach($args_types as $name => $type)
			if($type == 'unipath') {
				$uni_result = __uni_with_start_data(
					$tree[$lv-1]['data'], 
					isset($tree[$lv-1]['metadata']) 
						? $tree[$lv-1]['metadata'][0] 
						: $tree[$lv-1]['data_type'], 
					isset($tree[$lv-1]['data_tracking']) 
						? $tree[$lv-1]['data_tracking'] 
						: isset($tree[$lv-1]['metadata']) 
						? $tree[$lv-1]['metadata']
						: null, 
					$args[$name]);
				$args[$name] = $uni_result['data'];
			} 

		$result = array(
			'data' => $tree[$lv]['data'], 
			'data_type' => $tree[$lv]['metadata'][0],
			'metadata' => $tree[$lv]['metadata']);
			
		$func_name = strtolower(substr($cursor_arg1['name'], 0, $pos));
		switch($func_name) {
			// TODO content-type
			case 'bodyHtml': case 'bodyhtml':
				$result['data']['body'] = $args[0];
				break;
			case 'from':
/*				$result['data']['from'] = $args;
				break; */
			case 'to':
			case 'cc':
			case 'bcc':
				if(!isset($result['data'][$func_name]))
					$result['data'][$func_name] = array();
					
				// TODO сделать проверку на дубли
				if(is_array($args[0])) 
					foreach($args[0] as $email_rec) $result['data'][$func_name][] = $email_rec;
				else 
					$result['data'][$func_name][] = $args;
				break;
			case 'last_send_result':
			case 'last_error':
				if(array_key_exists($func_name, $tree[$lv]['metadata']))
					return array(
						'data' => $tree[$lv]['metadata'][$func_name],
						'data_type' => gettype($tree[$lv]['metadata'][$func_name]),
						'metadata' => array(gettype($tree[$lv]['metadata'][$func_name]), 'key()' => $func_name));
				else
					return array('data' => null, 'data_type' => 'null', 'metadata' => array('null', 'key()' => $func_name));
				break;
			case 'attacheFile': case 'attachefile':
			case 'addFile': case 'addfile':
			case 'file':
				if(!isset($result['data']['files']))
					$result['data']['files'] = array();
				if(is_array($args[0])) 
					foreach($args[0] as $file_rec) $result['data']['files'][] = $file_rec;
				elseif(isset($args['name']))
					$result['data']['files'][] = array($args[0], $args['name']);
				else
					$result['data']['files'][] = $args;
				break;
			case 'send':
				assert('!empty($tree[$lv]["data"]["to"]);');
// var_dump($tree[$lv-1]);
				$to = $tree[$lv]['data']['to'][0];
				$to = is_array($to) && count($to) > 1 
					? '=?UTF-8?B?'.base64_encode($to[1]).'?= <'.$to[0].'>'
					: $to[0];
				$subject = empty($tree[$lv]['data']['subject'])?'':'=?UTF-8?B?'.base64_encode($tree[$lv]['data']['subject']).'?=';
				$headers = _cursor_email($tree, $lv, 'eval', array('name' => 'resultHeaders()'));
// var_dump($headers);
				$body = _cursor_email($tree, $lv, 'eval', array('name' => 'resultMessage()'));

				$result['metadata']['last_send_result'] = call_user_func(
					isset($GLOBALS['unipath_test_mail_function']) ? 
						$GLOBALS['unipath_test_mail_function'] 
						: 'mail', 
					$to, $subject, $body['data'], 
					$headers['data']);
				$result['metadata']['last_error'] = error_get_last();
				break;
			default: // subject, body
				$result['data'][$func_name] = $args[0];
				break;
		}
// print_r($result);
		return $result;
	}
	
	// EVAL xxx
	elseif($cursor_cmd == 'eval') {
		if(array_key_exists($cursor_arg1['name'], $tree[$lv-1]['data']))
			return array(
			'data' => $tree[$lv-1]['data'][$cursor_arg1['name']],
			'data_type' => gettype($tree[$lv-1]['data'][$cursor_arg1['name']]),
			'metadata' => array(gettype($tree[$lv-1]['data'][$cursor_arg1['name']]), 'key()' => $cursor_arg1['name']));
		else
			return array('data' => null, 'data_type' => 'null', 'metadata' => array('null', 'key()' => $cursor_arg1['name']));
	}
}

function __test_mail($to, $subject, $message, $additional_headers = null, $additional_parameters = null) {
// var_dump(func_get_args());
	$GLOBALS['unipath_test_mail_arguments'] = func_get_args();
	return true;
}

function _tests_email() {
	$GLOBALS['unipath_test_mail_function'] = '__test_mail';
	
	file_put_contents('unipath.email.test_attache_file', '999');
	
// $GLOBALS['unipath_debug'] = true;
	$unipath = "/email()/to('root@localhost')/from('nobody@localhost')/subject(`test`)/body('123')/attacheFile(`unipath.email.test_attache_file`, name=`test.txt`))/send()/last_send_result()";
	echo "<h3>--- $unipath ---</h3>";
	$result = uni($unipath);
// print_r($result);
	assert('$result == true; /* '.var_export($result, true).' */');
	
// print_r($GLOBALS['unipath_test_mail_arguments']);
	assert('$GLOBALS[\'unipath_test_mail_arguments\'][0] == "root@localhost"; /* '.var_export($GLOBALS['unipath_test_mail_arguments'][0], true).' */');
	assert('$GLOBALS[\'unipath_test_mail_arguments\'][1] == "=?UTF-8?B?dGVzdA==?="; /* '.var_export($GLOBALS['unipath_test_mail_arguments'][1], true).' */');
	
	preg_match('~boundary="([0-9a-f]+)"~', $GLOBALS['unipath_test_mail_arguments'][3], $boundary);
	$expect = "From: nobody@localhost\r
MIME-Version: 1.0\r
Content-Type: multipart/mixed; boundary=\"{$boundary[1]}\"";
	$expect2 = "Content-Transfer-Encoding: 7bit\r
\r
This is a MIME encoded message.\r
\r
--{$boundary[1]}\r
Content-Type: text/html; charset=\"utf-8\"\r
Content-Transfer-Encoding: 8bit\r
\r
123\r
--{$boundary[1]}\r
Content-Type: application/octet-stream; name=\"test.txt\"\r
Content-Transfer-Encoding: base64\r
Content-Disposition: attachment; filename=\"test.txt\"\r
\r
OTk5\r
\r
--{$boundary[1]}--";

/*	for($i = 0; $i < strlen($expect); $i++) {
		var_dump($i, substr($expect, 0, $i) == substr($GLOBALS['unipath_test_mail_arguments'][3], 0, $i), substr($expect, 0, $i));
	} */

	assert('$GLOBALS[\'unipath_test_mail_arguments\'][3] == $expect; /* '.var_export($GLOBALS['unipath_test_mail_arguments'][3], true).' */');

	assert('$GLOBALS[\'unipath_test_mail_arguments\'][2] == $expect2; /* '.var_export($GLOBALS['unipath_test_mail_arguments'][2], true).' */');
}

}